interval/window.go

```diff
@@ -0,0 +1,310 @@
+// Window boundaries are defined to start at the epoch plus the offset.
+// Each subsequent window starts at a multiple of the every duration.
+// Each window's length is the start boundary plus the period.
+// Every must not be a mix of months and nanoseconds in order to preserve constant time bounds lookup.
+func NewWindow(every, period, offset values.Duration) (Window, error) {
+	zero := epoch.Add(offset)
+	w := Window{
+		every:      every,
+		period:     period,
+		zero:       zero,
+		zeroMonths: monthsSince(zero),
+	}
+	if err := w.isValid(); err != nil {
+		return Window{}, err
+	}
+	return w, nil
+}
+
+func (w Window) isValid() error {
+	if w.every.IsZero() {
+		return errors.New(codes.Invalid, "duration used as an interval cannot be zero")
+	}
+	if w.every.IsMixed() {
+		const docURL = "https://v2.docs.influxdata.com/v2.0/reference/flux/stdlib/built-in/transformations/window/#calendar-months-and-years"
+		return errors.New(codes.Invalid, "duration used as an interval cannot mix month and nanosecond units").
+			WithDocURL(docURL)
+	}
+	// TODO(nathanielc): what about negative every is that allowed?
+	if w.every.IsNegative() {
+		return errors.New(codes.Invalid, "duration used as an interval cannot be negative")
+	}
+	return nil
+}
+
+// GetLatestBounds returns the bounds for the latest window bounds
+// that contains the given time t.  For underlapping windows that
+// do not contain time t, the window directly before time t will be returned.
+func (w Window) GetLatestBounds(t values.Time) Bounds {
+	// Get the latest index that should contain the time t
+	index := w.lastIndex(t)
+	// Construct the bounds from the index
+	start := w.zero.Add(w.every.Mul(index))
+	b := Bounds{
+		start: start,
+		stop:  start.Add(w.period),
+		index: index,
+	}
+	// If the period is negative its possible future bounds can still contain this point
+	if w.period.IsNegative() {
+		// swap start and stop since the period was negative
+		b.start, b.stop = b.stop, b.start
+		// If period is NOT mixed we can do a direct calculation
+		// to determine how far into the future a bounds may be found.
+		if !w.period.IsMixed() {
+			// Since its not mixed we can adjust the index closer based
+			// on how many windows a period can span
+			var period, every int64
+			if w.every.MonthsOnly() {
+				every = w.every.Months()
+				period = w.period.Months()
+			} else {
+				every = w.every.Nanoseconds()
+				period = w.period.Nanoseconds()
+			}
+			if period > every {
+				indexDelta := period / every
+				index += int(indexDelta)
+			}
+		}
+		// Now do a direct search
+		next := w.NextBounds(b)
+		for next.Contains(t) {
+			b = next
+			next = w.NextBounds(next)
+		}
+	}
+	return b
+}
+
+// GetOverlappingBounds returns a slice of bounds that overlaps the input bounds.
+// The returned set of bounds are ordered by decreasing time.
+func (w Window) GetOverlappingBounds(start, stop values.Time) []Bounds {
```

Comment:
I think if we're rethinking the API for this I'd like to avoid this function. It's a very convenient function that generally leads a developer working on the standard library down a path that they will regret.